### PR TITLE
drivers: adc: Add Microchip MCP3914 SPI 8ch ADC driver

### PR DESCRIPTION
**Pull request by dean-at-evos**
New driver for the above chip on SPI bus. Runs in async, interrupt driven mode only. Basic chip features supported at this stage.

---

A new driver for the microchip mcp3914 on SPI bus. Tested on ESP32c3 with basic functionality.

### CODE REVIEW

**Comment by dean-at-evos** [comment]
Opening this up for a quick review. We do have a few things to add for our own app;

* per channel PGA_GAIN support (there is a TODO in comments where that might go)
* offset/slope correction register set
* _maybe_  CRC config/interrupt support

There are two things I would like to get feedback on early;

* The AFE on reset is 24 bit (BE) on the SPI bus, the driver is doing sign extension/cpu endian correction and outputs in `int32_t` and report resolution as 32 bit. Seems like the best place to do it to me .. is the driver expected to output device native data?
* For our app it is nice to know when a sample has occurred. I pass a timestamp down inside the driver but want to expose it. At the moment if there is a non-null user-data pointer in the sequence-options the driver treats it is an arrt of timestamps. Of course the can be documented, but it is the correct use of the user data field... or did I miss an obvious way to do this .. or something else??

File: drivers/adc/adc_mcp3914.c [hunk 1]
```diff
@@ -0,0 +1,847 @@
...
+		.osr = DT_STRING_UNQUOTED(DT_DRV_INST(inst), hware_osr),                           \
+		.clksrc = DT_STRING_UNQUOTED(DT_DRV_INST(inst), mclk_src)};                        \
+                                                                                                   \
+	static struct mcp3914_data mcp3914_data_##inst = {                                         \
+		ADC_CONTEXT_INIT_LOCK(mcp3914_data_##inst, ctx),                                   \
+		ADC_CONTEXT_INIT_SYNC(mcp3914_data_##inst, ctx),                                   \
+	};                                                                                         \
+                                                                                                   \
+	DEVICE_DT_INST_DEFINE(inst, &mcp3914_init, NULL, &mcp3914_data_##inst,                     \
+			      &mcp3914_config_##inst, POST_KERNEL, CONFIG_SPI_INIT_PRIORITY,       \
```
Comment (salmankhalid-cpi) [inline drivers/adc/adc_mcp3914.c:844]: Should not use same as SPI_INIT_PRIORITY as SPI needs to be initialized before MCP driver. Maybe make it configuration

**Comment by salmankhalid-cpi** [comment]
@dean-at-evos I also needed mcp3914 driver for zephyr found your pull request. I am testing it with STM32G491 (not able to successfully test it yet). Also added few comments.

**Review by salmankhalid-cpi — COMMENTED** [review]
Still working on review .. few initial comments.

**Comment by dean-at-evos** [comment]
> @dean-at-evos I also needed mcp3914 driver for zephyr found your pull request. I am testing it with STM32G491 (not able to successfully test it yet). Also added few comments.

Thanks for your interest.

TLDR;  I'll get to the review soon. I am on another project for a month or two, but I am coming back to this, Q2 24

You say you haven't been able to test it. Does that mean it doesn't [build|load|work] for you, or you just haven't had the [hardware|time] to do it? I have had a machine running with this driver (two chips) for a 16 channel system for a few months now, but that is just on one SOC. If you need help with implementing to get things running let me know, .... I'll look at the review soon

**Comment by dean-at-evos** [comment]
Apologies but this is no longer stale ... back in active development :)

**Comment by meuren** [comment]
At the end of 2025, Microchip announced that it intended to release a new version of the MCP391x, the MCP391xB. [Link](https://www.microchip.com/en-us/about/media-center/blog/2025/announcing-the-mcp391xb-series)
Is this driver still under development? Are there plans to expand the driver?

Best regards and many thanks
